### PR TITLE
Feature/hw 58098 php sdk transfer refunds

### PR DESCRIPTION
**Pull request by Rrathinasabapath**

### CODE REVIEW

**Review by ramahalingam — COMMENTED** [review]
Please delete all the files which are not required to be pushed and have only the files that are part of the sdk : Hyperwallet.php, TransferRefund.php and HyperwalletTest.php. Delete all the other files.

File: src/Hyperwallet/Hyperwallet.php [hunk 1]
```diff
@@ -420,6 +422,21 @@ public function createTransfer(Transfer $transfer) {
         return new Transfer($body);
     }
 
+    public function createTransferRefund($transferToken, TransferRefund $transferRefund = null) {
```
Comment (ramahalingam) [inline src/Hyperwallet/Hyperwallet.php:425]: TransferRefund $transferRefund = null can be just TransferRefund $transferRefund

File: src/Hyperwallet/Hyperwallet.php [hunk 2]
```diff
@@ -438,6 +455,20 @@ public function getTransfer($transferToken) {
         return new Transfer($body);
     }
 
+    public function getTransferRefund($transferToken,$refundToken) {
```
Comment (ramahalingam) [inline src/Hyperwallet/Hyperwallet.php:458]: ($transferToken,$refundToken). Pls give a space here.

File: src/Hyperwallet/Hyperwallet.php [hunk 3]
```diff
@@ -438,6 +455,20 @@ public function getTransfer($transferToken) {
...
+        if (empty($transferToken)) {
+            throw new HyperwalletArgumentException('transferToken is required!');
+        }
+        if (empty($refundToken)) {
+            throw new HyperwalletArgumentException('refundToken is required!');
+        }
+        $body = $this->client->doGet('/rest/v3/transfers/{transfer-token}/refunds/{refund-token}', array(
+                'transfer-token' => $transferToken,
+                'refund-token'=>$refundToken),
```
Comment (ramahalingam) [inline src/Hyperwallet/Hyperwallet.php:467]: Format pls

File: src/Hyperwallet/Hyperwallet.php [hunk 4]
```diff
@@ -453,6 +484,26 @@ public function listTransfers($options = array()) {
...
+        if (empty($transferToken)) {
+            throw new HyperwalletArgumentException('transferToken is required!');
+        }
+        $body = $this->client->doGet('/rest/v3/transfers/{transfer-token}/refunds', array('transfer-token' => $transferToken), $options);
+        return new ListResponse($body, function($entry) {
+            return new TransferRefund($entry);
+        });
+    }
+
+
```
Comment (ramahalingam) [inline src/Hyperwallet/Hyperwallet.php:505]: Delete extra lines here.

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 5]
```diff
@@ -13,24 +13,28 @@
...
 use Hyperwallet\Response\ErrorResponse;
 use Hyperwallet\Util\ApiClient;
 
-class HyperwalletTest extends \PHPUnit_Framework_TestCase {
+class HyperwalletTest extends \PHPUnit_Framework_TestCase
+{
+    private $includeIntegrationTest = false; //change this value to true if integration tests have to be run
 
-    public function testConstructor_throwErrorIfUsernameIsEmpty() {
+    public function testConstructor_throwErrorIfUsernameIsEmpty()
```
Comment (ramahalingam) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:36]: Revert the changes here. Pls revert all the formatting changes done in this whole file.

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 6]
```diff
@@ -3966,4 +4182,415 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+
+
+    }
+
+    public function testCreateTransferRefund_noTransferRefund(){
+
+        $userName = "test-username";
+        $password = "test-password";
+        $transferToken = "transferToken";
+            $client = new Hyperwallet($userName, $password);
```
Comment (ramahalingam) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4216]: format pls

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 7]
```diff
@@ -3966,4 +4182,415 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+            $this->assertEquals('transferRefund is required!', $e->getMessage());
+        }
+    }
+
+
+    public function testCreateTransferRefund_noTransferToken(){
+
+        $userName = "test-username";
+        $password = "test-password";
+        $clientRefundId = 6712348070812;
```
Comment (ramahalingam) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4231]: clientRefundId is a string. Pls change it.

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 8]
```diff
@@ -3966,4 +4182,415 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        $client = new Hyperwallet($userName, $password);
+
+        try {
+            $client->createTransferRefund(null, $transferRefund);
+            $this->fail('HyperwalletArgumentException expected');
+        } catch (HyperwalletArgumentException $e) {
+            $this->assertEquals('transferToken is required!', $e->getMessage());
+        }
+
+
```
Comment (ramahalingam) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4251]: Delete this extra line.

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 9]
```diff
@@ -3966,4 +4182,415 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        $password = "test-password";
+        $clientRefundId = 6712348070813;
+        $sourceAmount = 20.0;
+        $notes = "notes";
+        $memo = "memo";
+
+        $foreignExchange1 = array('sourceAmount'=>'200.00','sourceCurrency'=>'USD','destinationCurrency'=>'CAD',
+            'destinationAmount'=>'100.00','rate'=>'2.3');
+
+
```
Comment (ramahalingam) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4267]: Delete this extra line.

File: src/Hyperwallet/Hyperwallet.php (hunk 1 above)
Comment (Rrathinasabapath) [reply src/Hyperwallet/Hyperwallet.php:425]: This is to accept null $transferRefund value. However when null value is passed, proper user friendly error message is returned. Removing 'null' value fails the test.

File: src/Hyperwallet/Hyperwallet.php (hunk 3 above)
Comment (Rrathinasabapath) [reply src/Hyperwallet/Hyperwallet.php:467]: Done

File: src/Hyperwallet/Hyperwallet.php (hunk 4 above)
Comment (Rrathinasabapath) [reply src/Hyperwallet/Hyperwallet.php:505]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 5 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:36]: Done

**Review by ramahalingam — CHANGES_REQUESTED** [review]

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 6 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4216]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 7 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4231]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 8 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4251]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 9 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4267]: Done

File: src/Hyperwallet/Hyperwallet.php (hunk 1 above)
Comment (ramahalingam) [reply src/Hyperwallet/Hyperwallet.php:426]: I still feel that = null should be removed.

**Review by ramahalingam — CHANGES_REQUESTED** [review]
Delete the file .phpunit.result.cache from this PR

File: src/Hyperwallet/Hyperwallet.php (hunk 2 above)
Comment (Rrathinasabapath) [reply src/Hyperwallet/Hyperwallet.php]: Done

File: src/Hyperwallet/Hyperwallet.php (hunk 1 above)
Comment (Rrathinasabapath) [reply src/Hyperwallet/Hyperwallet.php:426]: Done.

**Review by ramahalingam — APPROVED** [review]

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 10]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        $client = new Hyperwallet($userName, $password);
+
+        try {
+            $client->createTransferRefund($transferToken, $transferRefund);
+            $this->fail('HyperwalletArgumentException expected');
+        } catch (HyperwalletArgumentException $e) {
+            $this->assertEquals('clientRefundId is required!', $e->getMessage());
+        }
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:3955]: Pls remove extra lines

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 11]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        try {
+            $client->createTransferRefund($transferToken, null);
+            $this->fail('HyperwalletArgumentException expected');
+        } catch (HyperwalletArgumentException $e) {
+            $this->assertEquals('transferRefund is required!', $e->getMessage());
+        }
+    }
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:3974]: Extra line

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 12]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        try {
+            $client->createTransferRefund(null, $transferRefund);
+            $this->fail('HyperwalletArgumentException expected');
+        } catch (HyperwalletArgumentException $e) {
+            $this->assertEquals('transferToken is required!', $e->getMessage());
+        }
+    }
+
+//******
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4000]: commented line above can be removed

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 13]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        $this->assertEquals($transferRefund->getMemo(), $memo);
+
+        $foreignExchange1 = $transferRefund->getForeignExchanges()[0];
+        $this->assertEquals($foreignExchange1['sourceAmount'], "200.00");
+        $this->assertEquals($foreignExchange1['sourceCurrency'], "USD");
+        $this->assertEquals($foreignExchange1['destinationAmount'], "100.00");
+        $this->assertEquals($foreignExchange1['destinationCurrency'], "CAD");
+        $this->assertEquals($foreignExchange1['rate'], "2.3");
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4043]: Extra line

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 14]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        // Validate mock
+        \Phake::verify($apiClientMock)->doPost('/rest/v3/transfers/{transfer-token}/refunds',
+            array('transfer-token' => $transferToken), $transferRefund, array());
+    }
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4049]: pls remove this line

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 15]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        }
+        $this->assertNotNull($transferRefund);
+        $this->assertEquals(array('token' => $refundToken), $transferRefund->getProperties());
+
+        // Validate mock
+        \Phake::verify($apiClientMock)->doGet('/rest/v3/transfers/{transfer-token}/refunds/{refund-token}',
+            $uriParams, $queryParams);
+    }
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4103]: pls remove this line

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 16]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+        // Run test
+        $transferRefundList = $client->listTransferRefunds($transferToken);
+        $this->assertEquals(array('sourceCurrency' => $sourceCurrency), $transferRefundList[0]->getProperties()[0]);
+        // Validate mock
+        \Phake::verify($apiClientMock)->doGet('/rest/v3/transfers/{transfer-token}/refunds',
+            $uriParams, $queryParams);
+    }
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4127]: same as above

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 17]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+            $transferRefundList = $hyperwallet->listTransferRefunds($transferToken);
+            var_dump('Transfer Refund list received', $transferRefundList);
+            echo "Listed transfer refund successfully";
+        } catch (\Hyperwallet\Exception\HyperwalletException $e) {
+            echo $e->getMessage();
+            die("\n");
+        }
+    }
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4231]: extra line

File: tests/Hyperwallet/Tests/HyperwalletTest.php [hunk 18]
```diff
@@ -3966,4 +3930,396 @@ public function testListVenmoAccountStatusTransitions_withParameters()
...
+                ->setDestinationAmount($destinationAmount)
+                ->setDestinationCurrency($destinationCurrency)
+                ->setNotes("Partial-Balance Transfer")
+                ->setMemo("TransferClientId56387")
+                ->setSourceToken($userToken)
+                ->setDestinationToken($destinationToken)
+            );
+            var_dump('Transfer created', $transfer);
+
+
```
Comment (arao6) [inline tests/Hyperwallet/Tests/HyperwalletTest.php:4307]: extra line

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 10 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:3955]: Done

Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:3955]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 11 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:3974]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 12 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4000]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 13 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4043]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 14 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4049]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 15 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4103]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 16 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4127]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 17 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4231]: Done

File: tests/Hyperwallet/Tests/HyperwalletTest.php (hunk 18 above)
Comment (Rrathinasabapath) [reply tests/Hyperwallet/Tests/HyperwalletTest.php:4307]: Done

**Review by arao6 — APPROVED** [review]

**Review by arao6 — APPROVED** [review]

**Review by peter-joseph — APPROVED** [review]
